File: forensics-btmgen/src/test/java/de/burger/forensics/plugin/engine/JavaPrefilterTest.java
// DEST: src/test/java/de/burger/forensics/plugin/engine/JavaPrefilterTest.java
package de.burger.forensics.plugin.engine;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import org.junit.jupiter.api.Test;

class JavaPrefilterTest {

    @Test
    void removesNoiseButPreservesStructure() {
        String src = String.join("\n",
                "package demo;",
                "// line comment with if (false) { fail(); }",
                "class A {",
                "    String s = \"keep \\\"quotes\\\" // not a comment\";",
                "    /* block",
                "       comment */",
                "    if (x == \"OK\") { /*inline*/ System.out.println(x); } // tail comment",
                "    char c = 'x';",
                "}");

        String prefiltered = JavaPrefilter.prefilterJava(src);

        assertFalse(prefiltered.contains("line comment"));
        assertFalse(prefiltered.contains("block\n                   comment"));
        assertFalse(prefiltered.contains("keep \"quotes\" // not a comment"));
        assertTrue(prefiltered.contains("class A {"));
        assertTrue(prefiltered.contains("if (x == \""));
        assertFalse(prefiltered.contains("'x'"));
        long prefilteredNewlines = prefiltered.chars().filter(ch -> ch == '\n').count();
        long originalNewlines = src.chars().filter(ch -> ch == '\n').count();
        assertEquals(originalNewlines, prefilteredNewlines,
                "Prefiltering should preserve newline structure");
    }
}

File: forensics-btmgen/src/test/kotlin/de/burger/forensics/plugin/engine/JavaPrefilterTest.kt
package de.burger.forensics.plugin.engine

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class JavaPrefilterTest {
    @Test
    fun `removes noise but preserves structure`() {
        val src = """
            package demo;
            // line comment with if (false) { fail(); }
            class A {
                String s = "keep \"quotes\" // not a comment";
                /* block
                   comment */
                if (x == "OK") { /*inline*/ System.out.println(x); } // tail comment
                char c = 'x';
            }
        """.trimIndent()

        val prefiltered = JavaPrefilter.prefilterJava(src)

        assertThat(prefiltered).doesNotContain("line comment")
        assertThat(prefiltered).doesNotContain("block\n                   comment")
        assertThat(prefiltered).doesNotContain("keep \"quotes\" // not a comment")
        assertThat(prefiltered).contains("class A {")
        assertThat(prefiltered).contains("if (x == \"")
        assertThat(prefiltered).doesNotContain("'x'")
        assertThat(prefiltered.count { it == '\n' }).isEqualTo(src.count { it == '\n' })
    }
}
